### PR TITLE
Support VisionEncoderDecoderModel and add an example

### PR DESCRIPTION
**Pull request by PenghuiCheng**
# What does this PR do?

This PR will support VisionEncoderDecoderModel like [TrOCR](https://huggingface.co/microsoft/trocr-small-handwritten)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by ftian1 — APPROVED** [review]

File: examples/optical-character-recognition/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Optical Character Recognition
+
+The script [`run_trocr.py`](https://github.com/huggingface/optimum/blob/main/examples/optical-character-recognition/run_trocr.py)
```
Comment (echarlaix) [inline examples/optical-character-recognition/README.md:19]: The path should be :
```suggestion
The script [`run_trocr.py`](https://github.com/huggingface/optimum-intel/blob/main/examples/neural_compressor/optical-character-recognition/run_trocr.py)
```

File: examples/optical-character-recognition/run_trocr.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+from datasets import load_metric
+from PIL import Image
+from torch.utils.data import DataLoader, Dataset
+from transformers import TrOCRProcessor, VisionEncoderDecoderModel
+
+from optimum.intel.neural_compressor import IncOptimizer, IncQuantizationConfig, IncQuantizationMode, IncQuantizer
+from optimum.intel.neural_compressor.quantization import IncQuantizedModelForVision2Seq
+
+
+parser = argparse.ArgumentParser(description="PyTorch ImageNet Training")
```
Comment (echarlaix) [inline examples/optical-character-recognition/run_trocr.py:16]: I think it would be nice to use `HfArgumentParser` and create classes such as `ModelArguments` and `OptimizationArguments` to stay consistent with the other examples scripts

File: optimum/intel/neural_compressor/quantization.py [hunk 3]
```diff
@@ -358,3 +359,8 @@ class IncQuantizedModelForMaskedLM(IncQuantizedModel):
 class IncQuantizedModelForXLNetLM(IncQuantizedModel):
 
     TRANSFORMERS_AUTO_CLASS = XLNetLMHeadModel
+
+
```
Comment (echarlaix) [inline optimum/intel/neural_compressor/quantization.py:363]: There seem to be an error in the documentation generation, could you rebase ?

File: examples/optical-character-recognition/run_trocr.py (hunk 2 above)
Comment (PenghuiCheng) [reply examples/optical-character-recognition/run_trocr.py]: done

File: examples/neural_compressor/optical-character-recognition/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Optical Character Recognition
+
+The script [`run_trocr.py`](https://github.com/huggingface/optimum-intel/blob/main/examples/neural_compressor/optical-character-recognition/run_trocr.py)
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/README.md:19]: I would rename the script to run_ocr.py in order to match what we currently have in the examples. And in the case we currently can only use TrOCR models I would specify it in the README

File: examples/neural_compressor/optical-character-recognition/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+allows us to apply different quantization approaches (such as dynamic, static and aware-training quantization) as well as pruning
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/README.md:20]:
```suggestion
allows us to apply different quantization approaches (such as dynamic and static quantization) 
```

File: examples/neural_compressor/optical-character-recognition/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    --model_name_or_path microsoft/trocr-small-handwritten \
+    --datasets_dir IAM \
+    --apply_quantization \
+    --quantization_approach static \
+    --verify_loading \
+    --output_dir /tmp/trocr_output
+```
+
+In order to apply dynamic, static or aware-training quantization, `quantization_approach` must be set to 
+respectively `dynamic`, `static` or `aware_training`.
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/README.md:44]:
```suggestion
In order to apply dynamic and static quantization, `quantization_approach` must be set to 
respectively `dynamic` or `static`.
```

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    Arguments pertaining to what type of optimization we are going to apply on the model.
+    """
+
+    apply_quantization: bool = field(
+        default=False,
+        metadata={"help": "Whether or not to apply quantization."},
+    )
+    quantization_approach: Optional[str] = field(
+        default=None,
+        metadata={"help": "Quantization approach. Supported approach are static, dynamic and aware_training."},
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:86]:
```suggestion
        metadata={"help": "Quantization approach. Supported approach are static and dynamic."},
```

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+            optim_args.quantization_config if optim_args.quantization_config is not None else default_config,
+            config_file_name="quantization.yml",
+            cache_dir=model_args.cache_dir,
+        )
+        # Set metric tolerance if specified
+        if optim_args.tolerance_criterion is not None:
+            q8_config.set_tolerance(optim_args.tolerance_criterion)
+        # Set quantization approach if specified
+        if optim_args.quantization_approach is not None:
+            supported_approach = {"static", "dynamic", "aware_training"}
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:274]:
```suggestion
            supported_approach = {"static", "dynamic"}
```

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    # make sure vocab size is set correctly
+    model.config.vocab_size = model.config.decoder.vocab_size
+
+    # set beam search parameters
+    model.config.eos_token_id = processor.tokenizer.sep_token_id
+    model.config.max_length = 64
+    model.config.early_stopping = True
+    model.config.no_repeat_ngram_size = 3
+    model.config.length_penalty = 2.0
+    model.config.num_beams = 4
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:194]: Could we add these arguments to specified `Arguments` such as https://github.com/huggingface/transformers/blob/main/examples/pytorch/translation/run_translation.py#L202 ?

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+
+            encoding = {"pixel_values": pixel_values.squeeze(), "labels": torch.tensor(labels)}
+            return encoding
+
+    processor = TrOCRProcessor.from_pretrained(model_args.model_name_or_path)
+    test_dataset = IAMDataset(root_dir=test_dir, df=df, processor=processor)
+
+    test_dataloader = DataLoader(test_dataset, batch_size=training_args.per_device_eval_batch_size)
+
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:180]: The device will always be cpu in our case so I would replace it with
```suggestion
    device = torch.device("cpu")
```

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+        # dynamic quantization will be added when torch FX is more mature
+        if quant_approach != IncQuantizationMode.DYNAMIC:
+            q8_config.set_config("model.framework", "pytorch_fx")
+        q8_config.set_config("tuning.accuracy_criterion.higher_is_better", False)
+        quantizer = IncQuantizer(q8_config, eval_func=eval_func, calib_dataloader=test_dataloader)
+        optimizer = IncOptimizer(model, quantizer=quantizer)
+        q_model = optimizer.fit()
+        optimizer.save_pretrained(training_args.output_dir)
+
+    if optim_args.verify_loading:
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:292]: To verify the quantized model loading, I would add an evaluation step in order to compare it with the one of `q_model` (that what we currently do in the other examples scripts) which could also be replaced with inference. If not we should remove the `verify_loading` flag in my opinion.

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+        torch.backends.quantized.engine = "onednn"
+        print("loading int8 model...")
+        model = IncQuantizedModelForVision2Seq.from_pretrained(training_args.output_dir)
+        model.eval()
+
+    print("Running evaluation...")
+    final_score = eval_func(model, 20)
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:299]:
```suggestion
    final_score = eval_func(q_model, 20)
```

File: examples/neural_compressor/optical-character-recognition/run_trocr.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+    )
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    test_file = os.path.join(data_args.datasets_dir, "gt_test.txt")
+    test_dir = os.path.join(data_args.datasets_dir, "image")
+    df = pd.read_fwf(test_file, header=None)
+    df.rename(columns={0: "file_name", 1: "text"}, inplace=True)
+    del df[2]
+    df.head()
+
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_trocr.py:146]: We should add `data_args.max_calibration_samples` and `data_args.max_eval_samples` to control the number of calibration and evaluation samples

File: examples/neural_compressor/optical-character-recognition/README.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+## Run Command
+The following example applies post-training static quantization on TrOCR small fine-tuned on the IAM dataset.
+```bash
+python run_trocr.py \
+    --model_name_or_path microsoft/trocr-small-handwritten \
+    --datasets_dir IAM \
+    --apply_quantization \
+    --quantization_approach static \
+    --verify_loading \
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/README.md:39]: Do we need to set a `tolerance_criterion` value in order to not do multiple quantizartion passes ? (which could take some time)

**Review by echarlaix — COMMENTED** [review]
Thanks a lot for this addition, it looks great !

File: examples/neural_compressor/optical-character-recognition/README.md (hunk 14 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/README.md:40]: Hi, Ella, I set max_eval_samples to save the tuning time. Is it OK?

File: examples/neural_compressor/optical-character-recognition/README.md (hunk 4 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/README.md]: Done

File: examples/neural_compressor/optical-character-recognition/run_trocr.py (hunk 9 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/run_trocr.py]: Done

File: examples/neural_compressor/optical-character-recognition/run_trocr.py (hunk 13 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/run_trocr.py]: Done

File: examples/neural_compressor/optical-character-recognition/run_ocr.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+            print("Latency: %.3f ms" % (batch_time.avg * 1000))
+        print("Throughput: %.3f images/sec" % (training_args.per_device_eval_batch_size / batch_time.avg))
+
+        # TODO: this should also be done with the ProgressMeter
+        print("cer: {score:.5f}".format(score=(score)))
+
+        return score
+
+    if optim_args.apply_quantization:
+        torch.backends.quantized.engine = "onednn"
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_ocr.py:275]: Could you remove this line in order for all users to be able to run this script ? We currently have no dependency to onednn

**Review by echarlaix — COMMENTED** [review]
Thanks for iterating on this PR !

File: examples/neural_compressor/optical-character-recognition/README.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+## Run Command
+The following example applies post-training static quantization on TrOCR small fine-tuned on the IAM dataset.
+```bash
+python run_ocr.py \
+    --model_name_or_path microsoft/trocr-small-handwritten \
+    --datasets_dir IAM \
+    --apply_quantization \
+    --quantization_approach static \
+    --max_eval_samples 1000 \
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/README.md:39]: Would it be reasonable to reduce the number of `max_eval_samples` ? (to something like 500), otherwise evaluation can take quite some time

File: examples/neural_compressor/optical-character-recognition/README.md (hunk 14 above)
Comment (echarlaix) [reply examples/neural_compressor/optical-character-recognition/README.md:40]: Thanks for adding `max_eval_samples`  it's very useful. I think we can still add a `tolerance_criterion` in the README (in order for user to not have to wait too much) but I don't have a strong opinion on this so I am ok with keeping it like this if you prefer it this way

File: examples/neural_compressor/optical-character-recognition/run_ocr.py [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+            # add batch to metric
+            cer.add_batch(predictions=pred_str, references=label_str)
+            if iters is not None and i >= iters:
+                break
+        score = cer.compute()
+        print("Batch size = %d" % training_args.per_device_eval_batch_size)
+        if training_args.per_device_eval_batch_size == 1:
+            print("Latency: %.3f ms" % (batch_time.avg * 1000))
+        print("Throughput: %.3f images/sec" % (training_args.per_device_eval_batch_size / batch_time.avg))
```
[the remark this answers is not in the document]
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/run_ocr.py:267]: Yes, you are right. maybe we can add a conditional judgment before this.

File: examples/neural_compressor/optical-character-recognition/README.md (hunk 14 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/README.md:41]: Yes, I had added max_eval_samples and tolerance_criterion.  Thanks!

File: examples/neural_compressor/optical-character-recognition/README.md (hunk 16 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/README.md]: Had set to 500.

File: examples/neural_compressor/optical-character-recognition/run_ocr.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+        # dynamic quantization will be added when torch FX is more mature
+        if quant_approach != IncQuantizationMode.DYNAMIC:
+            q8_config.set_config("model.framework", "pytorch_fx")
+        if quant_approach == IncQuantizationMode.STATIC:
+            q8_config.set_config("quantization.calibration.sampling_size", [data_args.max_calibration_samples])
+        q8_config.set_config("tuning.accuracy_criterion.higher_is_better", False)
+        quantizer = IncQuantizer(q8_config, eval_func=eval_func, calib_dataloader=test_dataloader)
+        optimizer = IncOptimizer(model, quantizer=quantizer)
+        q_model = optimizer.fit()
+        torch.backends.quantized.engine = "onednn"
```
Comment (echarlaix) [inline examples/neural_compressor/optical-character-recognition/run_ocr.py:308]: It would be great to remove this line as well (for the same reason as the previous one)

File: examples/neural_compressor/optical-character-recognition/run_ocr.py (hunk 15 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/optical-character-recognition/run_ocr.py]: removed

**Review by echarlaix — COMMENTED** [review]
Thanks a lot for this addition @PenghuiCheng 🔥Before we merge, can you run [make style](https://github.com/huggingface/optimum-intel/blob/main/Makefile#L27) on your branch to make sure the quality test passes ?

File: examples/neural_compressor/optical-character-recognition/run_ocr.py (hunk 18 above)
Comment (echarlaix) [reply examples/neural_compressor/optical-character-recognition/run_ocr.py]: Thank you !

**Review by echarlaix — APPROVED** [review]
LGTM, thanks for this addition !
